core/store/src/trie/mem/mod.rs

```diff
@@ -1,8 +1,206 @@
+        self.roots.get(state_root).map(|ids| ids[0].as_ptr(self.arena.memory()))
+    }
+
+    /// Expires all trie roots corresponding to a height smaller than
+    /// `block_height`. This internally manages refcounts. If a trie root
+    /// is expired but is still used at a higher height, it will still be
+    /// valid until all references to that root expires.
+    pub fn delete_until_height(&mut self, block_height: BlockHeight) {
+        let mut to_delete = vec![];
+        self.heights.retain(|height, state_roots| {
+            if *height < block_height {
+                for state_root in state_roots {
+                    to_delete.push(*state_root)
+                }
+                false
+            } else {
+                true
+            }
+        });
+        for state_root in to_delete {
+            self.delete_root(&state_root);
+        }
+    }
+
+    fn delete_root(&mut self, state_root: &CryptoHash) {
+        if let Some(ids) = self.roots.get_mut(state_root) {
+            let last_id = ids.last().unwrap();
+            let new_ref = last_id.remove_ref(&mut self.arena);
+            if new_ref == 0 {
+                ids.pop();
+                if ids.is_empty() {
+                    self.roots.remove(state_root);
+                }
+            }
+        } else {
+            debug_assert!(false, "Deleting non-existent root: {}", state_root);
+        }
+        MEM_TRIE_NUM_ROOTS
+            .with_label_values(&[&self.shard_uid.shard_id.to_string()])
+            .set(self.roots.len() as i64);
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::node::{InputMemTrieNode, MemTrieNodeId};
+    use super::MemTries;
+    use crate::NibbleSlice;
+    use near_primitives::hash::CryptoHash;
+    use near_primitives::shard_layout::ShardUId;
+    use near_primitives::state::FlatStateValue;
+    use near_primitives::types::BlockHeight;
+    use rand::seq::SliceRandom;
+    use rand::Rng;
+
+    #[test]
+    fn test_refcount() {
+        // Here we test multiple cases:
+        //  - Each height possibly having multiple state roots (due to forks)
+        //    (and possibly with the same state roots)
+        //  - Each state root possibly having multiple actual nodes that have
+        //    the same hash (as we don't deduplicate in general)
+        //  - A state root being possibly the same as another of a different
+        //    height.
+        //
+        // And we make sure that the GC refcounting works correctly.
+        let mut tries = MemTries::new(1024 * 1024 * 1024, ShardUId::single_shard());
+        let mut available_hashes: Vec<(BlockHeight, CryptoHash)> = Vec::new();
+        for height in 100..=200 {
+            let num_roots_at_height = rand::thread_rng().gen_range(1..=4);
+            for _ in 0..num_roots_at_height {
+                match rand::thread_rng().gen_range(0..4) {
+                    0 | 1 | 2 | 3 if available_hashes.is_empty() => {
```

Comment:
So I basically want it so that when it's 0, 1, 2, make a new hash. If it's 3, reuse an existing hash. Except that if I don't have any hashes I can't trigger the existing hash case, thus a conditional here.

Reply:
Yeah, but there is `0 | 1 | 2 | 3` so it's always "make a new hash"?
Also, it was not clear what is the purpose of the number. I thought that it is shard id because we have 4 shards right now.
Can we mention it somewhere? Maybe make it self-descriptive with enum?
```
enum AddingRootMode {
  NewRoot // <- 0..3
  OldRoot // <- 3
}
```

Reply:
it's 0, or 1, or 2, or (3 if empty).

I flipped it around to (0 if not empty) and (else). Is this clearer now? With an enum I'd have to use a weighted sampling which is more complicated.

Reply:
> it's 0, or 1, or 2, or (3 if empty).

Oh, I didn't know that "if" has priority over "|". Yeah, now it is clearer, thank you!